ocaml/xapi/xapi_pgpu.ml

```diff
@@ -68,15 +68,62 @@ let sync_pci_hidden ~__context ~pgpu ~pci =
     | `enabled | `disable_on_reboot -> false
   end
 
-let update_gpus ~__context ~host =
+(* If this PCI device is by AMD and if (ON THE HOST RUNNING THIS CODE) it has
+ * a valid-looking phys_fn symlink in its entry under /sys/bus/pci/...
+ * then:
+ *   assume it is a Virtual Function (we should already have checked it is a display device)
+ *   set the physical_function field in its DB object to point to its PF PCI object
+ *   return true meaning it is a VF (even if we couldn't find a PF PCI object for it)
+ * else return false *)
+(* This is not in the .mli *)
+let mxgpu_set_phys_fn_ref ~__context pci =
+  Xapi_pci.int_of_id (Db.PCI.get_vendor_id ~__context ~self:pci) = Xapi_vgpu_type.Vendor_amd.vendor_id &&
+  Db.PCI.get_virtual_functions ~__context ~self:pci = [] && (* i.e. we don't already know it is a phys fn *)
+  (* A better name for pci_id would be pci_address. *)
+  let pci_addr = (Db.PCI.get_pci_id ~__context ~self:pci) in
+  (* E.g. path = "/sys/bus/pci/devices/0000:88:00.0/" *)
+  let path = Printf.sprintf "/sys/bus/pci/devices/%s/physfn" pci_addr in
+  try (
+    (* No problem if there's no such symlink: we'll handle the exception. *)
+    let linktarget = Unix.readlink path in
+    (* Expect linktarget to look like "../0000:8c:01.0" *)
+    (* If it does then cut off "../" and look up Db ref with that pci_id (i.e. address), and create a link in DB *)
+    if (try Scanf.sscanf linktarget "../%_4x:%_2x:%_2x:%_1x%!" true
```

Comment:
I think it would be safer to not make any assumptions about the path other than the basename. The kernel or driver, or even `readlink`, might change such that an absolute path is returned, or one that looks slightly different in some other way. We could instead find the PCI address by `Filename.basename`.